yjit/src/codegen.rs

```diff
@@ -2210,51 +2211,48 @@ fn gen_checktype(
             | (RUBY_T_HASH, Type::Hash) => {
                 // guaranteed type match
                 let stack_ret = ctx.stack_push(Type::True);
-                mov(cb, stack_ret, uimm_opnd(Qtrue.as_u64()));
+                asm.mov(stack_ret, Opnd::UImm(Qtrue.as_u64()));
                 return KeepCompiling;
             }
             _ if val_type.is_imm() || val_type.is_specific() => {
                 // guaranteed not to match T_STRING/T_ARRAY/T_HASH
                 let stack_ret = ctx.stack_push(Type::False);
-                mov(cb, stack_ret, uimm_opnd(Qfalse.as_u64()));
+                asm.mov(stack_ret, Opnd::UImm(Qfalse.as_u64()));
                 return KeepCompiling;
             }
             _ => (),
         }
 
-        mov(cb, REG0, val);
-        mov(cb, REG1, uimm_opnd(Qfalse.as_u64()));
-
-        let ret = cb.new_label("ret".to_string());
+        let ret = asm.new_label("ret");
+        let ret_opnd = Opnd::UImm(Qfalse.as_u64());
```

Comment:
Maybe we should just replace references to `ret_opnd` with Qfalse since it seems to be constant?

Reply:
🤔 The reason for assigning it was to match the `cmov` that we make in `gen_checktype` after the type check:
```
        mov(cb, REG1, uimm_opnd(Qfalse.as_u64()));
[...]
        mov(cb, REG0, mem_opnd(64, REG0, RUBY_OFFSET_RBASIC_FLAGS));
        and(cb, REG0, uimm_opnd(RUBY_T_MASK as u64));
        cmp(cb, REG0, uimm_opnd(type_val as u64));
        mov(cb, REG0, uimm_opnd(Qtrue.as_u64()));
        // REG1 contains Qfalse from above
        cmove(cb, REG1, REG0);
        
[...]

        mov(cb, stack_ret, REG1);
```

I could look into doing it differently if that doesn't make sense ?

Reply:
Oh I didn't realize I could do that, I'll give it a shot thanks.

Reply:
Actually this is making me realize that it should be:
`let ret_opnd = asm.csel_e(Qfalse.into(), Qtrue.into());`

We use 3-address instructions instead of 2-address now.